crates/workspace/src/pane.rs

```diff
@@ -1761,30 +1774,64 @@ impl Pane {
                         );
 
                     if let Some(entry) = single_entry_to_resolve {
-                        let parent_abs_path = pane
+                        let paths = pane
                             .update(cx, |pane, cx| {
                                 pane.workspace.update(cx, |workspace, cx| {
                                     let project = workspace.project().read(cx);
                                     project.worktree_for_entry(entry, cx).and_then(|worktree| {
                                         let worktree = worktree.read(cx);
                                         let entry = worktree.entry_for_id(entry)?;
-                                        let abs_path = worktree.absolutize(&entry.path).ok()?;
-                                        let parent = if entry.is_symlink {
-                                            abs_path.canonicalize().ok()?
-                                        } else {
-                                            abs_path
-                                        }
-                                        .parent()?
-                                        .to_path_buf();
-                                        Some(parent)
+                                        let rel_path = &entry.path;
+                                        let abs_path = (|| {
+                                            let abs_path = worktree.absolutize(rel_path).ok()?;
+                                            let parent = if entry.is_symlink {
+                                                abs_path.canonicalize().ok()?
+                                            } else {
+                                                abs_path
+                                            }
+                                            .to_path_buf();
+                                            Some(parent)
+                                        })();
+                                        Some((Arc::clone(rel_path), abs_path))
                                     })
                                 })
                             })
                             .ok()
                             .flatten();
 
+                        let (rel_path, abs_path) =
```

Comment:
All these lines below is a very dirty code, I liked the previous version more.

Why do we have to drag the `entry.path` out of all this while `if let Some(entry) = single_entry_to_resolve` is the source of the very same relative path?

Cannot we just leave the `let parent_abs_path = ...` exactly as it was before and get the relative path out of the entry?

Reply:
This is partially because I couldn't figure out how to get `entry.path` out of it, so I tried to use tuple as the return type, and later deconstruct it, and create new `parent_abs_path` for the `Open In Terminal` to use.

Though even if we revert those lines, I can think of no way to `absolute_path` and `relative_path` out of the clouser. And the `entry` is not the same since it's being redeclared by `entry = worktree.entry_for_id(entry)?;`
so in order to get the absolute path, I have to repeat those lines.

I think the better solution is to return a tuple of relative_path, absolute_path, and parent_abs_path from the clouser and use them

```rust
let (abs_path, rel_path, parent_abs_path) = pane
    .update(cx, |pane, cx| {
        pane.workspace.update(cx, |workspace, cx| {
            let project = workspace.project().read(cx);
            project.worktree_for_entry(entry, cx).and_then(|worktree| {
                let worktree = worktree.read(cx);
                let entry = worktree.entry_for_id(entry)?;
                let abs_path = worktree.absolutize(&entry.path).ok()?;

                let abs_path = Some(abs_path.to_path_buf());
                let rel_path = entry.path.to_path_buf();

                let parent = if entry.is_symlink {
                    abs_path.canonicalize().ok()?
                } else {
                    abs_path
                }
                .parent()?
                .to_path_buf();

                Some((abs_path, rel_path, parent))
            })
        })
    })
    .ok()
    .flatten()
    .map(|(abs_path, rel_path, parent)| (abs_path, rel_path, parent));
```

Reply:
I think we can leave that block alone, it is quite packed already.

Why cannot you extract a common method for retrieving an absolute path for a given entry? Then you could do `.and_then(|abs_path| abs_path|.parent())` on it in the original block, and use it right away in the new function.

Reply:
Or just repeat the abs path extraction in your separate function for handling the "copy path" action?

Reply:
I was afraid of repeating a pre-existing block into my own function, but if I can extract and use, then...

<details>
<summary>For readability</summary>

```rust
...
let absolute_path = Pane::get_absolute_path(&pane, cx, entry); 
let parent_abs_path = Pane::get_absolute_path(&pane, cx, entry)
.and_then(|abs_path| Some(abs_path.parent()?.to_path_buf()));
...
```
In place of that `quite packed` block inside `render_tab`

The new (or old) get_absolute_path function
```rust
fn get_absolute_path(
        pane: &View<Pane>,
        cx: &mut WindowContext,
        entry: ProjectEntryId,
    ) -> Option<PathBuf> {
        pane.update(cx, |pane, cx| {
            pane.workspace.update(cx, |workspace, cx| {
                let project = workspace.project().read(cx);
                project.worktree_for_entry(entry, cx).and_then(|worktree| {
                    let worktree = worktree.read(cx);
                    let entry = worktree.entry_for_id(entry)?;
                    let abs_path = worktree.absolutize(&entry.path).ok()?;
                    let path = if entry.is_symlink {
                        abs_path.canonicalize().ok()?
                    } else {
                        abs_path
                    };

                    Some(path)
                })
            })
        })
        .ok()
        .flatten()
    }
```

Old `copy_relative_path`
```rust
fn copy_relative_path(&mut self, _: &CopyRelativePath, cx: &mut ViewContext<Self>) {
        if let Some(clipboard_text) = self
            .active_item()
            .as_ref()
            .and_then(|entry| entry.project_path(cx))
            .map(|p| p.path.to_string_lossy().to_string())
        {
            cx.write_to_clipboard(ClipboardItem::new(clipboard_text));
        }
    }
```

And finally the menu entries themselves,
```rust
...
menu = menu
    .separator()
    .when_some(absolute_path, |menu, path| {
        menu.entry(
            "Copy Path",
            Some(Box::new(CopyPath)),
            cx.handler_for(&pane, move |_, cx| {
                cx.write_to_clipboard(ClipboardItem::new(
                    path.to_string_lossy().to_string(),
                ));
            }),
        )
    })
    .entry(
        "Copy Relative Path",
        Some(Box::new(CopyRelativePath)),
        cx.handler_for(&pane, move |pane, cx| {
            pane.copy_relative_path(&CopyRelativePath, cx);
        }),
    )
    ...
```

</details>

Reply:
I hope this time it is all right, I have had compilation errors quite a few times now. Learnt a lot of rust by refactoring it out 😓.
Sorry for this many re-reviews